Simplified/SignInLogic/NYPLSignInBusinessLogic+CardCreation.swift

```diff
@@ -9,10 +9,115 @@
+                                "libraryAccountName": libraryAccount?.name ?? "N/A",
+      ])
+      completion(nil, error)
+      self.cardCreationIsOngoing = false
+      self.cardCreationLock.unlock()
+      return
+    }
+
+    // Verify if the native card creator is supported for this library,
+    // otherwise default to web
+    guard libraryAccount?.details?.supportsCardCreator ?? false else {
+      let title = NSLocalizedString("eCard",
+                                    comment: "Title for web-based card creator page")
+      let msg = NSLocalizedString("The page could not load due to a connection error.",
+                                  comment: "Message for errors loading a HTML page")
+      let webVC = RemoteHTMLViewController(URL: signUpURL,
+                                           title: title,
+                                           failureMessage: msg)
+      completion(UINavigationController(rootViewController: webVC), nil)
+      self.cardCreationIsOngoing = false
+      self.cardCreationLock.unlock()
+      return
+    }
+    
+    guard let coordinator = makeCardCreationCoordinator() else {
+      // This should only happen when CardCreator credentials decode failed (which is very unlikely to happen)
+      // and the errors are already logged when we retrieve the credentials
+      let description = NSLocalizedString("Sign up process fails to initiate.", comment: "Message describing the CardCreator flow failed to be initiated")
+      let error = NSError(domain: NYPLErrorLogger.clientDomain,
+                          code: NYPLErrorCode.cardCreatorCredentialsDecodeFail.rawValue,
+                          userInfo: [
+                            NSLocalizedDescriptionKey: description
+                          ])
+      completion(nil, error)
+      self.cardCreationIsOngoing = false
+      self.cardCreationLock.unlock()
+      return
+    }
+    
+    cardCreationCoordinator = coordinator
+    
+    coordinator.configuration.completionHandler = { [weak self] username, pin, isUserInitiated in
+      guard let self = self else {
+        return
+      }
 
+      if isUserInitiated {
+        // Dismiss CardCreator when user finishes Credential Review
+        self.uiDelegate?.dismiss(animated: true, completion: nil)
+      } else {
+        if let usernameTextField = self.uiDelegate?.usernameTextField, let PINTextField = self.uiDelegate?.PINTextField {
+          usernameTextField.text = username
+          PINTextField.text = pin
+        }
+        self.isLoggingInAfterSignUp = true
+        self.logIn()
+      }
+      self.cardCreationCoordinator = nil
+    }
+    
+    coordinator.startRegularFlow { [weak self] result in
+      switch result {
+      case .success(let navVC):
+        completion(navVC, nil)
+      case .fail(let error):
+        NYPLErrorLogger.logError(error, summary: "Regular Card Creation error")
+        completion(nil, error)
+        self?.cardCreationCoordinator = nil
```

Comment:
We eventually do this when the whole CardCreator flow completes, on line 104. I did some testing and result is we do need the coordinator even after initiating the flow, otherwise the API requests fail. I can't really tell whats happening without doing some in-depth investigation, I am guessing we lost the reference to the `ISSOToken` since we don't hold any reference to the view controllers.

Reply:
ah ok i that case no worries.